src/apify/scrapy/extensions/_httpcache.py

```diff
@@ -0,0 +1,179 @@
+from __future__ import annotations
+
+import gzip
+import io
+import pickle
+import struct
+from logging import getLogger
+from time import time
+from typing import TYPE_CHECKING
+
+from scrapy.http.headers import Headers
+from scrapy.responsetypes import responsetypes
+
+from apify import Configuration
+from apify.apify_storage_client import ApifyStorageClient
+from apify.scrapy._async_thread import AsyncThread
+from apify.storages import KeyValueStore
+
+if TYPE_CHECKING:
+    from scrapy import Request, Spider
+    from scrapy.http.response import Response
+    from scrapy.settings import BaseSettings
+    from scrapy.utils.request import RequestFingerprinterProtocol
+
+logger = getLogger(__name__)
+
+
+class ApifyCacheStorage:
+    """A Scrapy cache storage that uses the Apify `KeyValueStore` to store responses.
+
+    It can be set as a storage for Scrapy's built-in `HttpCacheMiddleware`, which caches
+    responses to requests. See HTTPCache middleware settings (prefixed with `HTTPCACHE_`)
+    in the Scrapy documentation for more information. Requires the asyncio Twisted reactor
+    to be installed.
+    """
+
+    def __init__(self, settings: BaseSettings) -> None:
+        self._expiration_max_items = 100
+        self._expiration_secs: int = settings.getint('HTTPCACHE_EXPIRATION_SECS')
+        self._spider: Spider | None = None
+        self._kvs: KeyValueStore | None = None
+        self._fingerprinter: RequestFingerprinterProtocol | None = None
+        self._async_thread: AsyncThread | None = None
+
+    def open_spider(self, spider: Spider) -> None:
+        """Open the cache storage for a spider."""
+        logger.debug('Using Apify key value cache storage', extra={'spider': spider})
+        self._spider = spider
+        self._fingerprinter = spider.crawler.request_fingerprinter
+        kvs_name = f'httpcache-{spider.name}'
```

Comment:
It looks like the KVS name cannot contain underscores.

Reply:
Ha! My spiders apparently always had names with dashes. Can I bring in a dependency on `python-slugify`? Or maybe that's an overkill? I'll check what are allowed characters for spiders in Scrapy. Maybe some sort of replace would suffice.

Reply:
Please do not use any other 3rd party library, that is not currently installed.

Reply:
Addressed in [4ff56d7](https://github.com/apify/apify-sdk-python/pull/403/commits/4ff56d7ebc57ec9cfb22cf201276a8e63b8e6d74). I opted for simplicity - the KVS name is determined from the spider name in a naive way, but I think it should suffice to most people. One thing I want to keep is to be able to inspect the KVS in the Console and immediately know to which scraper it belongs:

<img width="1101" alt="Screenshot 2025-03-18 at 9 27 31" src="https://github.com/user-attachments/assets/0a8d47d5-4e57-4909-a07f-62070fe8d68c" />

I don't anticipate that people have spiders which result in duplicate KVS names. If they do, we can hash the original spider name and append a piece of that hash to the KVS name, but for now that felt like an overkill. Let me know what you think.

Reply:
Just a note, I've read the Scrapy docs and the spider name doesn't seem to be limited in any way to certain sets of characters. It's just an arbitrary unicode string. If we later find issues in my implementation, I'd opt for bringing `python-slugify` in, because I wouldn't like to re-implement it here, that library is battle-tested.

Reply:
This is good enough, thanks.